examples/models/burgers_one2n.py

```diff
@@ -0,0 +1,276 @@
+        dt = Constant(parameters.timestep)
+        nu = parameters.viscosity(mesh)
+
+        # Define variational formulation
+        V = self.function_space
+        u = Function(V)
+        u_ = Function(V)
+        v = TestFunction(V)
+        self._form = (
+            inner((u - u_) / dt, v) * dx
+            + inner(dot(u, nabla_grad(u)), v) * dx
+            + nu * inner(grad(u), grad(v)) * dx
+        )
+        problem = NonlinearVariationalProblem(self._form, u)
+
+        # Set initial condition
+        u_.project(parameters.ic(mesh))
+
+        # Create solver
+        self._solver = NonlinearVariationalSolver(problem)
+        self._solution = u
+
+    @property
+    def function_space(self):
+        r"""
+        The :math:`\mathbb P2` finite element space.
+        """
+        return get_function_space(self.mesh)
+
+    @property
+    def form(self):
+        """
+        The weak form of Burgers equation
+        """
+        return self._form
+
+    @property
+    def solution(self):
+        return self._solution
+
+    def iterate(self, **kwargs):
+        """
+        Take a single timestep of Burgers equation
+        """
+        self._solver.solve()
+        
+        
+class Solver_one2n(nn_adapt.solving.Solver):
+    """
+    Solver object based on current mesh and state.
+    """
+
+    def __init__(self, mesh, ic, **kwargs):
+        """
+        :arg mesh: the mesh to define the solver on
+        :arg ic: the current state / initial condition
+        """
+        self.mesh = mesh
+
+        # Collect parameters
+        self.tt_steps = parameters.tt_steps
+        dt = Constant(parameters.timestep)
+        
+        # Physical parameters
+        nu = parameters.viscosity(mesh)
+        self.nu = nu
+        
+        # Define variational formulation
+        V = self.function_space
+        self.u = Function(V)
+        self.u_ = Function(V)
+        self.v = TestFunction(V)
+        
+        self._form = (
+            inner((self.u - self.u_) / dt, self.v) * dx
+            + inner(dot(self.u, nabla_grad(self.u)), self.v) * dx
+            + nu * inner(grad(self.u), grad(self.v)) * dx
+        )
+        
+        # Define initial conditions
+        ic = parameters.ic(self.mesh)
+        self.u.project(ic)
+        
+        # Set solutions
+        self._solutions = []
+
+    @property
+    def function_space(self):
+        r"""
+        The :math:`\mathbb P2` finite element space.
+        """
+        return get_function_space(self.mesh)
+
+    @property
+    def form(self):
+        """
+        The weak form of Burgers equation
+        """
+        return self._form
+
+    @property
+    def solution(self):
+        return self._solutions
+    
+    @property
+    def adj_solution(self):
+        return self._adj_solution
+    
+    def adjoint_iteration(self):
+        """
+        Get the forward solutions of Burgers equation
+        """
+        J_form = inner(self.u, self.u)*ds(2)
```

Comment:
It would be good if you could use the output of the `get_qoi` function, rather than hard-coding the QoI. If you do that, then the code will be more flexible, e.g. we can make a different QoI choice.